cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,327 @@
+            raise DataSimulatorError("Specify a positive number of cassettes.")
+        if size_of_cassette <= 0 or not isinstance(size_of_cassette, int):
+            raise DataSimulatorError(
+                "Specify a positive number of cut-sites" " per cassette."
+            )
+
+        self.size_of_cassette = size_of_cassette
+        self.number_of_cassettes = number_of_cassettes
+
+        if isinstance(mutation_rate, float):
+            if mutation_rate < 0:
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            number_of_characters = size_of_cassette * number_of_cassettes
+            self.mutation_rate_per_character = [
+                mutation_rate
+            ] * number_of_characters
+        else:
+            if len(mutation_rate) != (
+                self.number_of_cassettes * self.size_of_cassette
+            ):
+                raise DataSimulatorError(
+                    "Length of mutation rate array is not"
+                    " the same as the number of characters."
+                )
+
+            if np.any(np.array(mutation_rate) < 0):
+                raise DataSimulatorError(
+                    "Mutation rate needs to be" " non-negative."
+                )
+            self.mutation_rate_per_character = mutation_rate
+
+        self.mutation_priors = mutation_priors
+        if self.mutation_priors is None:
+            self.mutation_priors = {}
+            probabilites = [
+                state_distribution() for _ in range(number_of_states)
+            ]
+            Z = np.sum(probabilites)
+            for i in range(number_of_states):
+                self.mutation_priors[i + 1] = probabilites[i] / Z
+        else:
+            Z = np.sum([v for v in self.mutation_priors.values()])
+            if not math.isclose(Z, 1.0):
+                raise DataSimulatorError("Mutation priors do not sum to 1.")
+
+        self.heritable_silencing_rate = heritable_silencing_rate
+        self.stochastic_silencing_rate = stochastic_silencing_rate
+
+        self.random_seed = random_seed
+
+    def overlay_data(self, tree: CassiopeiaTree):
+        """Overlays Cas9-based lineage tracing data onto the CassiopeiaTree.
+
+        Args:
+            tree: Input CassiopeiaTree
+        """
+
+        if self.random_seed is not None:
+            np.random.seed(self.random_seed)
+
+        number_of_characters = self.number_of_cassettes * self.size_of_cassette
+
+        # initialize character states
+        character_matrix = {}
+        for node in tree.nodes:
+            character_matrix[node] = [-1] * number_of_characters
+
+        for node in tree.depth_first_traverse_nodes(tree.root, postorder=False):
+
+            if tree.is_root(node):
+                character_matrix[node] = [0] * number_of_characters
+                continue
+
+            parent = tree.parent(node)
```

Comment:
Sorry, I don't see how this can happen -- I'm checking specifically above whether or not the node is the root. You should never hit `parent = tree.parent(node)` if the node is the root.

Reply:
I missed this! Thank you for clarifying